File: src/lib/sanctions.ts
import type { UsingClient } from "seyfert";
import { config } from "../config";
import { sanctionRequests, sanctions } from "./db";
import type { SanctionRequest, SanctionRequestType } from "./types";

export interface SubmitRequestInput {
	type: SanctionRequestType;
	targetId: string;
	targetTag: string;
	reason: string;
	evidence: string | null;
	submittedBy: string;
	submittedByTag: string;
	guildId: string;
}

export async function submitRequest(input: SubmitRequestInput): Promise<SanctionRequest> {
	return sanctionRequests.create({
		...input,
		status: "pending",
		requiredApprovals: config.requiredApprovals,
		approvals: [],
		declines: [],
		declineReason: null,
		submittedAt: new Date().toISOString(),
		resolvedAt: null,
	});
}

export type VoteDecision = "approve" | "decline";

export interface VoteResult {
	request: SanctionRequest;
	/** Whether this vote just caused the request to resolve (approved/declined). */
	justResolved: boolean;
	enforcement?: EnforcementSummary;
}

export class VoteError extends Error {}

export async function castVote(
	client: UsingClient,
	requestId: string,
	reviewerId: string,
	decision: VoteDecision,
	declineReason?: string,
): Promise<VoteResult> {
	const request = await sanctionRequests.get(requestId);
	if (!request) throw new VoteError(`No submission found with id \`${requestId}\`.`);
	if (request.status !== "pending") {
		throw new VoteError(`That submission was already **${request.status}**.`);
	}

	const approvals = new Set(request.approvals);
	const declines = new Set(request.declines);

	if (decision === "approve") {
		declines.delete(reviewerId);
		approvals.add(reviewerId);
	} else {
		approvals.delete(reviewerId);
		declines.add(reviewerId);
	}

	const patch: Partial<SanctionRequest> = {
		approvals: [...approvals],
		declines: [...declines],
	};

	let justResolved = false;
	let enforcement: EnforcementSummary | undefined;

	if (approvals.size >= request.requiredApprovals) {
		patch.status = "approved";
		patch.resolvedAt = new Date().toISOString();
		justResolved = true;
	} else if (declines.size >= request.requiredApprovals) {
		patch.status = "declined";
		patch.resolvedAt = new Date().toISOString();
		patch.declineReason = declineReason ?? request.declineReason ?? null;
		justResolved = true;
	}

	const updated = await sanctionRequests.update(requestId, patch);

	if (justResolved && updated.status === "approved") {
		enforcement = await applyApprovedRequest(client, updated);
	}

	if (justResolved) {
		client.logger.info(
			`[nationseal] AUDIT submission=${updated.id} resolved=${updated.status} reviewers=${updated.approvals.length + updated.declines.length}`,
		);
	}

	return { request: updated, justResolved, enforcement };
}

export interface EnforcementSummary {
	guildsAttempted: number;
	guildsSucceeded: number;
}

/** Applies an approved submission: updates the shared list and enforces it live. */
async function applyApprovedRequest(
	client: UsingClient,
	request: SanctionRequest,
): Promise<EnforcementSummary> {
	if (request.type === "ban") {
		await sanctions.upsert(request.targetId, {
			reason: request.reason,
			requestId: request.id,
			addedBy: request.submittedBy,
			addedAt: new Date().toISOString(),
		});
		return banAcrossGuilds(client, request.targetId, `NationSeal sanction: ${request.reason}`);
	}

	await sanctions.remove(request.targetId);
	return unbanAcrossGuilds(client, request.targetId, "NationSeal sanction appeal approved");
}

/**
 * Returns the set of guild IDs the bot can currently see, while refusing to
 * ever ban the bot itself, the application owner, or any NationSeal owner.
 */
function isProtectedUser(userId: string, client: UsingClient): boolean {
	if (userId === client.botId) return true;
	if (config.ownerIds.includes(userId)) return true;
	return false;
}

async function getCachedGuildIds(client: UsingClient): Promise<string[]> {
	// Use the public `values()` API so we get fully-typed Guild objects with a
	// canonical `.id` field instead of scraping the cache's internal
	// `"<namespace>.<id>"` key encoding.
	const values = (await client.cache.guilds?.values()) ?? [];
	return values
		.map((guild) => guild?.id)
		.filter((id): id is string => typeof id === "string" && id.length > 0);
}

async function banSingleUserInGuild(
	client: UsingClient,
	guildId: string,
	userId: string,
	reason: string,
): Promise<boolean> {
	if (isProtectedUser(userId, client)) {
		client.logger.warn(`[nationseal] Refusing to ban protected user ${userId} in ${guildId}.`);
		return false;
	}
	try {
		await client.bans.create(
			guildId,
			userId,
			config.banDeleteMessageSeconds > 0
				? { delete_message_seconds: config.banDeleteMessageSeconds }
				: {},
			reason,
		);
		return true;
	} catch (error) {
		client.logger.warn(`[nationseal] Could not ban ${userId} in guild ${guildId}:`, error);
		return false;
	}
}

export async function banAcrossGuilds(
	client: UsingClient,
	userId: string,
	reason: string,
): Promise<EnforcementSummary> {
	const guildIds = await getCachedGuildIds(client);
	let succeeded = 0;

	for (const guildId of guildIds) {
		if (await banSingleUserInGuild(client, guildId, userId, reason)) {
			succeeded++;
		}
	}

	return { guildsAttempted: guildIds.length, guildsSucceeded: succeeded };
}

export async function unbanAcrossGuilds(
	client: UsingClient,
	userId: string,
	reason: string,
): Promise<EnforcementSummary> {
	const guildIds = await getCachedGuildIds(client);
	let succeeded = 0;

	for (const guildId of guildIds) {
		try {
			await client.bans.remove(guildId, userId, reason);
			succeeded++;
		} catch (error) {
			client.logger.warn(`[nationseal] Could not unban ${userId} in guild ${guildId}:`, error);
		}
	}

	return { guildsAttempted: guildIds.length, guildsSucceeded: succeeded };
}

// Per-guild in-flight guards so that a burst of `guildCreate` events (e.g. a
// bot shard reconnect, or a server farm inviting the bot many times in a
// short window) can't trigger overlapping ban syncs in the same guild.
const inFlightGuildSyncs = new Set<string>();

/** Applies every currently active sanction to a single guild, used when the bot joins a new server. */
export async function enforceAllSanctionsOnGuild(
	client: UsingClient,
	guildId: string,
): Promise<{ banned: number; total: number }> {
	if (inFlightGuildSyncs.has(guildId)) {
		client.logger.info(
			`[nationseal] Skipping duplicate guild sync for ${guildId}; already in progress.`,
		);
		return { banned: 0, total: 0 };
	}
	inFlightGuildSyncs.add(guildId);

	try {
		const allActive = await sanctions.listAll();
		if (allActive.length === 0) return { banned: 0, total: 0 };

		// Never include protected users (the bot itself, owners) in a bulk
		// ban sync — Discord would reject these anyway, but skipping them
		// here also keeps the audit log clean.
		const botId = client.botId;
		const eligible = allActive.filter((s) => !isProtectedUser(s.id, client) && s.id !== botId);
		if (eligible.length === 0) return { banned: 0, total: allActive.length };

		const BATCH_SIZE = 200;
		let banned = 0;

		for (let i = 0; i < eligible.length; i += BATCH_SIZE) {
			const batch = eligible.slice(i, i + BATCH_SIZE);
			try {
				const result = await client.bans.bulkCreate(
					guildId,
					{
						user_ids: batch.map((sanction) => sanction.id),
						...(config.banDeleteMessageSeconds > 0
							? { delete_message_seconds: config.banDeleteMessageSeconds }
							: {}),
					},
					"NationSeal: syncing shared sanction list",
				);
				banned += result.banned_users.length;
			} catch (error) {
				client.logger.warn(
					`[nationseal] Could not bulk-sync sanctions to guild ${guildId}:`,
					error,
				);
			}
		}

		client.logger.info(
			`[nationseal] AUDIT guildSync guild=${guildId} synced=${banned}/${eligible.length} (skipped ${allActive.length - eligible.length} protected)`,
		);

		return { banned, total: allActive.length };
	} finally {
		inFlightGuildSyncs.delete(guildId);
	}
}

File: nationseal/sanctions.py
"""Core multi-sig voting and cross-guild ban enforcement."""

from __future__ import annotations

import logging
from datetime import UTC, datetime
from typing import TYPE_CHECKING, cast

import discord

from nationseal.config import config
from nationseal.db import sanction_requests, sanctions
from nationseal.models import Sanction

if TYPE_CHECKING:
	from discord import Client

	from nationseal.models import SanctionRequest

logger = logging.getLogger("nationseal")


class VoteError(Exception):
	pass


class EnforcementSummary:
	def __init__(self) -> None:
		self.guilds_attempted = 0
		self.guilds_succeeded = 0


def _now_iso() -> str:
	return datetime.now(UTC).isoformat()


async def submit_request(input_: dict[str, object]) -> SanctionRequest:
	return await sanction_requests.create(
		cast(
			SanctionRequest,
			{
				**input_,
				"status": "pending",
				"requiredApprovals": config.required_approvals,
				"approvals": [],
				"declines": [],
				"declineReason": None,
				"submittedAt": _now_iso(),
				"resolvedAt": None,
			},
		)
	)


async def cast_vote(
	client: Client,
	request_id: str,
	reviewer_id: int | str,
	decision: str,
	decline_reason: str | None = None,
) -> dict[str, object]:
	reviewer_id_str = str(reviewer_id)
	request = await sanction_requests.get(request_id)
	if not request:
		raise VoteError(f"No submission found with id `{request_id}`.")
	if request["status"] != "pending":
		raise VoteError(f"That submission was already **{request['status']}**.")

	approvals = set(request["approvals"])
	declines = set(request["declines"])

	if decision == "approve":
		declines.discard(reviewer_id_str)
		approvals.add(reviewer_id_str)
	else:
		approvals.discard(reviewer_id_str)
		declines.add(reviewer_id_str)

	patch: dict[str, object] = {
		"approvals": list(approvals),
		"declines": list(declines),
	}

	just_resolved = False
	enforcement: EnforcementSummary | None = None

	if len(approvals) >= request["requiredApprovals"]:
		patch["status"] = "approved"
		patch["resolvedAt"] = _now_iso()
		just_resolved = True
	elif len(declines) >= request["requiredApprovals"]:
		patch["status"] = "declined"
		patch["resolvedAt"] = _now_iso()
		patch["declineReason"] = decline_reason or request.get("declineReason") or None
		just_resolved = True

	updated = await sanction_requests.update(request_id, patch)

	if just_resolved and updated["status"] == "approved":
		enforcement = await _apply_approved_request(client, updated)

	if just_resolved:
		logger.info(
			"[nationseal] AUDIT submission=%s resolved=%s reviewers=%d",
			updated["id"],
			updated["status"],
			len(updated["approvals"]) + len(updated["declines"]),
		)

	return {
		"request": updated,
		"justResolved": just_resolved,
		"enforcement": enforcement,
	}


async def _apply_approved_request(client: Client, request: SanctionRequest) -> EnforcementSummary:
	if request["type"] == "ban":
		await sanctions.upsert(
			request["targetId"],
			cast(
				Sanction,
				{
					"id": request["targetId"],
					"reason": request["reason"],
					"requestId": request["id"],
					"addedBy": request["submittedBy"],
					"addedAt": _now_iso(),
				},
			),
		)
		return await ban_across_guilds(client, request["targetId"], f"NationSeal sanction: {request['reason']}")

	await sanctions.remove(request["targetId"])
	return await unban_across_guilds(client, request["targetId"], "NationSeal sanction appeal approved")


def _is_protected_user(user_id: int | str, client: Client) -> bool:
	uid = str(user_id)
	if client.user and uid == str(client.user.id):
		return True
	return uid in config.owner_ids


def _get_cached_guild_ids(client: Client) -> list[int]:
	return [guild.id for guild in client.guilds]


async def _ban_single_user_in_guild(
	client: Client, guild_id: int, user_id: int, reason: str
) -> bool:
	if _is_protected_user(user_id, client):
		logger.warning("[nationseal] Refusing to ban protected user %s in %s.", user_id, guild_id)
		return False

	guild = client.get_guild(guild_id)
	if guild is None:
		return False

	try:
		kwargs: dict[str, object] = {}
		if config.ban_delete_message_seconds > 0:
			kwargs["delete_message_seconds"] = config.ban_delete_message_seconds
		if config.ban_delete_message_seconds > 0:
			await guild.ban(
				discord.Object(id=user_id),
				reason=reason,
				delete_message_seconds=config.ban_delete_message_seconds,
			)
		else:
			await guild.ban(discord.Object(id=user_id), reason=reason)
		return True
	except Exception as exc:  # noqa: BLE001
		logger.warning("[nationseal] Could not ban %s in guild %s: %s", user_id, guild_id, exc)
		return False


async def ban_across_guilds(client: Client, user_id: int | str, reason: str) -> EnforcementSummary:
	uid = int(user_id)
	guild_ids = _get_cached_guild_ids(client)
	summary = EnforcementSummary()

	for guild_id in guild_ids:
		summary.guilds_attempted += 1
		if await _ban_single_user_in_guild(client, guild_id, uid, reason):
			summary.guilds_succeeded += 1

	return summary


async def unban_across_guilds(client: Client, user_id: int | str, reason: str) -> EnforcementSummary:
	uid = int(user_id)
	guild_ids = _get_cached_guild_ids(client)
	summary = EnforcementSummary()

	for guild_id in guild_ids:
		guild = client.get_guild(guild_id)
		if guild is None:
			continue
		summary.guilds_attempted += 1
		try:
			await guild.unban(discord.Object(id=uid), reason=reason)
			summary.guilds_succeeded += 1
		except Exception as exc:  # noqa: BLE001
			logger.warning("[nationseal] Could not unban %s in guild %s: %s", user_id, guild_id, exc)

	return summary


_in_flight_guild_syncs: set[int] = set()


async def enforce_all_sanctions_on_guild(client: Client, guild_id: int | str) -> dict[str, int]:
	gid = int(guild_id)
	if gid in _in_flight_guild_syncs:
		logger.info("[nationseal] Skipping duplicate guild sync for %s; already in progress.", gid)
		return {"banned": 0, "total": 0}

	_in_flight_guild_syncs.add(gid)
	try:
		all_active = await sanctions.list_all()
		if not all_active:
			return {"banned": 0, "total": 0}

		eligible = [s for s in all_active if not _is_protected_user(s["id"], client)]
		if not eligible:
			return {"banned": 0, "total": len(all_active)}

		guild = client.get_guild(gid)
		if guild is None:
			return {"banned": 0, "total": len(all_active)}

		banned = 0
		batch_size = 200
		for i in range(0, len(eligible), batch_size):
			batch = eligible[i : i + batch_size]
			try:
				result = await guild.bulk_ban(
					[discord.Object(id=int(s["id"])) for s in batch],
					reason="NationSeal: syncing shared sanction list",
					delete_message_seconds=config.ban_delete_message_seconds or 86400,
				)
				banned += len(result.banned_users)  # type: ignore
			except Exception as exc:  # noqa: BLE001
				logger.warning("[nationseal] Could not bulk-sync sanctions to guild %s: %s", gid, exc)

		logger.info(
			"[nationseal] AUDIT guildSync guild=%s synced=%d/%d (skipped %d protected)",
			gid,
			banned,
			len(eligible),
			len(all_active) - len(eligible),
		)

		return {"banned": banned, "total": len(all_active)}
	finally:
		_in_flight_guild_syncs.discard(gid)
